fix the build issue

[chatelao](https://github.com/chatelao)
[micropython-tang-nano](https://github.com/chatelao/micropython-tang-nano)
Repository navigation
[Code](https://github.com/chatelao/micropython-tang-nano)
Issues
1
 (1)
Pull requests
10
 (10)
[Actions](https://github.com/chatelao/micropython-tang-nano/actions)
[Projects](https://github.com/chatelao/micropython-tang-nano/projects)
[Wiki](https://github.com/chatelao/micropython-tang-nano/wiki)
[Security](https://github.com/chatelao/micropython-tang-nano/security)
[Insights](https://github.com/chatelao/micropython-tang-nano/pulse)
[Settings](https://github.com/chatelao/micropython-tang-nano/settings)
[Build and Release MicroPython firmware](https://github.com/chatelao/micropython-tang-nano/actions/workflows/build.yml)
Merge pull request #139 from chatelao/release-firmware-and-bitstream-… #234
All jobs
Run details
Annotations
1 error and 1 warning
[build](https://github.com/chatelao/micropython-tang-nano/actions/runs/23458100470/job/68252559898#logs)
failed 38 minutes ago in 1m 21s
Search logs
2s
1s
32s
2s
15s
13s
13s
Run make -C src/ports/tang_nano_4k/ BUILD=build_hw
make: Entering directory '/home/runner/work/micropython-tang-nano/micropython-tang-nano/src/ports/tang_nano_4k'
Use make V=1 or set BUILD_VERBOSE in your environment to increase build verbosity.
mkdir -p build_hw/genhdr
GEN build_hw/genhdr/mpversion.h
GEN build_hw/genhdr/qstr.i.last
GEN build_hw/genhdr/qstr.split
GEN build_hw/genhdr/qstrdefs.collected.h
QSTR updated
GEN build_hw/genhdr/qstrdefs.generated.h
GEN build_hw/genhdr/moduledefs.split
GEN build_hw/genhdr/moduledefs.collected
Module registrations updated
GEN build_hw/genhdr/moduledefs.h
GEN build_hw/genhdr/root_pointers.split
GEN build_hw/genhdr/root_pointers.collected
Root pointer registrations updated
GEN build_hw/genhdr/root_pointers.h
GEN build_hw/genhdr/compressed.split
GEN build_hw/genhdr/compressed.collected
Compressed data updated
GEN build_hw/genhdr/compressed.data.h
mkdir -p build_hw/drivers/bus/
mkdir -p build_hw/extmod/
mkdir -p build_hw/lib/littlefs/
mkdir -p build_hw/py/
mkdir -p build_hw/shared/libc/
mkdir -p build_hw/shared/readline/
mkdir -p build_hw/shared/runtime/
mkdir -p build_hw/shared/timeutils/
CC ../../lib/micropython/py/mpstate.c
CC ../../lib/micropython/py/nlr.c
CC ../../lib/micropython/py/nlrx86.c
CC ../../lib/micropython/py/nlrx64.c
CC ../../lib/micropython/py/nlrthumb.c
CC ../../lib/micropython/py/nlraarch64.c
CC ../../lib/micropython/py/nlrmips.c
CC ../../lib/micropython/py/nlrpowerpc.c
CC ../../lib/micropython/py/nlrxtensa.c
CC ../../lib/micropython/py/nlrrv32.c
CC ../../lib/micropython/py/nlrrv64.c
CC ../../lib/micropython/py/nlrsetjmp.c
CC ../../lib/micropython/py/malloc.c
CC ../../lib/micropython/py/gc.c
CC ../../lib/micropython/py/pystack.c
CC ../../lib/micropython/py/qstr.c
CC ../../lib/micropython/py/vstr.c
CC ../../lib/micropython/py/mpprint.c
CC ../../lib/micropython/py/unicode.c
CC ../../lib/micropython/py/mpz.c
CC ../../lib/micropython/py/reader.c
CC ../../lib/micropython/py/lexer.c
CC ../../lib/micropython/py/parse.c
CC ../../lib/micropython/py/scope.c
CC ../../lib/micropython/py/compile.c
CC ../../lib/micropython/py/emitcommon.c
CC ../../lib/micropython/py/emitbc.c
CC ../../lib/micropython/py/asmbase.c
CC ../../lib/micropython/py/asmx64.c
CC ../../lib/micropython/py/emitnx64.c
CC ../../lib/micropython/py/asmx86.c
CC ../../lib/micropython/py/emitnx86.c
CC ../../lib/micropython/py/asmthumb.c
CC ../../lib/micropython/py/emitnthumb.c
CC ../../lib/micropython/py/emitinlinethumb.c
CC ../../lib/micropython/py/asmarm.c
CC ../../lib/micropython/py/emitnarm.c
CC ../../lib/micropython/py/asmxtensa.c
CC ../../lib/micropython/py/emitnxtensa.c
CC ../../lib/micropython/py/emitinlinextensa.c
CC ../../lib/micropython/py/emitnxtensawin.c
CC ../../lib/micropython/py/asmrv32.c
CC ../../lib/micropython/py/emitnrv32.c
CC ../../lib/micropython/py/emitndebug.c
CC ../../lib/micropython/py/formatfloat.c
CC ../../lib/micropython/py/parsenumbase.c
CC ../../lib/micropython/py/parsenum.c
CC ../../lib/micropython/py/emitglue.c
CC ../../lib/micropython/py/persistentcode.c
CC ../../lib/micropython/py/runtime.c
CC ../../lib/micropython/py/runtime_utils.c
CC ../../lib/micropython/py/scheduler.c
CC ../../lib/micropython/py/nativeglue.c
CC ../../lib/micropython/py/pairheap.c
CC ../../lib/micropython/py/ringbuf.c
CC ../../lib/micropython/py/cstack.c
CC ../../lib/micropython/py/stackctrl.c
CC ../../lib/micropython/py/argcheck.c
CC ../../lib/micropython/py/warning.c
CC ../../lib/micropython/py/profile.c
CC ../../lib/micropython/py/map.c
CC ../../lib/micropython/py/obj.c
CC ../../lib/micropython/py/objarray.c
CC ../../lib/micropython/py/objattrtuple.c
CC ../../lib/micropython/py/objbool.c
CC ../../lib/micropython/py/objboundmeth.c
CC ../../lib/micropython/py/objcell.c
CC ../../lib/micropython/py/objclosure.c
CC ../../lib/micropython/py/objcomplex.c
CC ../../lib/micropython/py/objdeque.c
CC ../../lib/micropython/py/objdict.c
CC ../../lib/micropython/py/objenumerate.c
CC ../../lib/micropython/py/objexcept.c
CC ../../lib/micropython/py/objfilter.c
CC ../../lib/micropython/py/objfloat.c
CC ../../lib/micropython/py/objfun.c
CC ../../lib/micropython/py/objgenerator.c
CC ../../lib/micropython/py/objgetitemiter.c
CC ../../lib/micropython/py/objint.c
CC ../../lib/micropython/py/objint_longlong.c
CC ../../lib/micropython/py/objint_mpz.c
CC ../../lib/micropython/py/objlist.c
CC ../../lib/micropython/py/objmap.c
CC ../../lib/micropython/py/objmodule.c
CC ../../lib/micropython/py/objobject.c
CC ../../lib/micropython/py/objpolyiter.c
CC ../../lib/micropython/py/objproperty.c
CC ../../lib/micropython/py/objnone.c
CC ../../lib/micropython/py/objnamedtuple.c
CC ../../lib/micropython/py/objrange.c
CC ../../lib/micropython/py/objreversed.c
CC ../../lib/micropython/py/objringio.c
CC ../../lib/micropython/py/objset.c
CC ../../lib/micropython/py/objsingleton.c
CC ../../lib/micropython/py/objslice.c
CC ../../lib/micropython/py/objstr.c
CC ../../lib/micropython/py/objstrunicode.c
CC ../../lib/micropython/py/objstringio.c
CC ../../lib/micropython/py/objtuple.c
CC ../../lib/micropython/py/objtype.c
CC ../../lib/micropython/py/objzip.c
CC ../../lib/micropython/py/opmethods.c
CC ../../lib/micropython/py/sequence.c
CC ../../lib/micropython/py/stream.c
CC ../../lib/micropython/py/binary.c
CC ../../lib/micropython/py/builtinimport.c
CC ../../lib/micropython/py/builtinevex.c
CC ../../lib/micropython/py/builtinhelp.c
CC ../../lib/micropython/py/modarray.c
CC ../../lib/micropython/py/modbuiltins.c
CC ../../lib/micropython/py/modcollections.c
CC ../../lib/micropython/py/modgc.c
CC ../../lib/micropython/py/modio.c
CC ../../lib/micropython/py/modmath.c
CC ../../lib/micropython/py/modcmath.c
CC ../../lib/micropython/py/modmicropython.c
CC ../../lib/micropython/py/modstruct.c
CC ../../lib/micropython/py/modsys.c
CC ../../lib/micropython/py/moderrno.c
CC ../../lib/micropython/py/modthread.c
CC ../../lib/micropython/py/vm.c
CC ../../lib/micropython/py/bc.c
CC ../../lib/micropython/py/showbc.c
CC ../../lib/micropython/py/repl.c
CC ../../lib/micropython/py/smallint.c
CC ../../lib/micropython/py/frozenmod.c
CC main.c
CC mphalport.c
CC uart.c
CC pin.c
CC timer.c
CC pwm.c
CC adc.c
CC flash.c
CC i2c.c
CC spi.c
CC rtc.c
CC wdt.c
CC modmachine.c
CC ../../lib/micropython/shared/timeutils/timeutils.c
CC ../../lib/micropython/extmod/modtime.c
CC ../../lib/micropython/extmod/machine_mem.c
CC ../../lib/micropython/extmod/machine_i2c.c
CC ../../lib/micropython/extmod/machine_spi.c
CC ../../lib/micropython/drivers/bus/softspi.c
CC ../../lib/micropython/extmod/virtpin.c
CC ../../lib/micropython/extmod/modvfs.c
CC ../../lib/micropython/extmod/vfs.c
CC ../../lib/micropython/extmod/vfs_lfs.c
CC ../../lib/micropython/extmod/vfs_reader.c
CC ../../lib/micropython/extmod/vfs_blockdev.c
CC ../../lib/micropython/extmod/modos.c
CC ../../lib/micropython/lib/littlefs/lfs2.c
CC ../../lib/micropython/lib/littlefs/lfs2_util.c
CC ../../lib/micropython/shared/libc/printf.c
CC ../../lib/micropython/shared/runtime/stdout_helpers.c
CC ../../lib/micropython/shared/readline/readline.c
CC ../../lib/micropython/shared/runtime/pyexec.c
LINK build_hw/firmware.elf
/usr/lib/gcc/arm-none-eabi/13.2.1/../../../arm-none-eabi/bin/ld: /usr/lib/gcc/arm-none-eabi/13.2.1/../../../arm-none-eabi/lib/thumb/v7-m/nofp/libc.a(libc_a-closer.o): in function `_close_r':
/build/newlib-38V0JC/newlib-4.4.0.20231231/build/arm-none-eabi/thumb/v7-m/nofp/newlib/../../../../../../newlib/libc/reent/closer.c:47:(.text+0xc): warning: _close is not implemented and will always fail
/usr/lib/gcc/arm-none-eabi/13.2.1/../../../arm-none-eabi/bin/ld: /usr/lib/gcc/arm-none-eabi/13.2.1/../../../arm-none-eabi/lib/thumb/v7-m/nofp/libc.a(libc_a-lseekr.o): in function `_lseek_r':
/build/newlib-38V0JC/newlib-4.4.0.20231231/build/arm-none-eabi/thumb/v7-m/nofp/newlib/../../../../../../newlib/libc/reent/lseekr.c:49:(.text+0x14): warning: _lseek is not implemented and will always fail
/usr/lib/gcc/arm-none-eabi/13.2.1/../../../arm-none-eabi/bin/ld: /usr/lib/gcc/arm-none-eabi/13.2.1/../../../arm-none-eabi/lib/thumb/v7-m/nofp/libc.a(libc_a-closer.o): note: the message above does not take linker garbage collection into account
/usr/lib/gcc/arm-none-eabi/13.2.1/../../../arm-none-eabi/bin/ld: /usr/lib/gcc/arm-none-eabi/13.2.1/../../../arm-none-eabi/lib/thumb/v7-m/nofp/libc.a(libc_a-readr.o): in function `_read_r':
/build/newlib-38V0JC/newlib-4.4.0.20231231/build/arm-none-eabi/thumb/v7-m/nofp/newlib/../../../../../../newlib/libc/reent/readr.c:49:(.text+0x14): warning: _read is not implemented and will always fail
/usr/lib/gcc/arm-none-eabi/13.2.1/../../../arm-none-eabi/bin/ld: /usr/lib/gcc/arm-none-eabi/13.2.1/../../../arm-none-eabi/lib/thumb/v7-m/nofp/libc.a(libc_a-writer.o): in function `_write_r':
/build/newlib-38V0JC/newlib-4.4.0.20231231/build/arm-none-eabi/thumb/v7-m/nofp/newlib/../../../../../../newlib/libc/reent/writer.c:49:(.text+0x14): warning: _write is not implemented and will always fail
/usr/lib/gcc/arm-none-eabi/13.2.1/../../../arm-none-eabi/bin/ld: /usr/lib/gcc/arm-none-eabi/13.2.1/../../../arm-none-eabi/lib/thumb/v7-m/nofp/libc.a(libc_a-lseekr.o): note: the message above does not take linker garbage collection into account
/usr/lib/gcc/arm-none-eabi/13.2.1/../../../arm-none-eabi/bin/ld: /usr/lib/gcc/arm-none-eabi/13.2.1/../../../arm-none-eabi/lib/thumb/v7-m/nofp/libc.a(libc_a-readr.o): note: the message above does not take linker garbage collection into account
/usr/lib/gcc/arm-none-eabi/13.2.1/../../../arm-none-eabi/bin/ld: build_hw/firmware.elf section `.text' will not fit in region `FLASH'
/usr/lib/gcc/arm-none-eabi/13.2.1/../../../arm-none-eabi/bin/ld: /usr/lib/gcc/arm-none-eabi/13.2.1/../../../arm-none-eabi/lib/thumb/v7-m/nofp/libc.a(libc_a-writer.o): note: the message above does not take linker garbage collection into account
/usr/lib/gcc/arm-none-eabi/13.2.1/../../../arm-none-eabi/bin/ld: region `FLASH' overflowed by 73636 bytes
collect2: error: ld returned 1 exit status
make: *** [Makefile:103: build_hw/firmware.elf] Error 1
make: Leaving directory '/home/runner/work/micropython-tang-nano/micropython-tang-nano/src/ports/tang_nano_4k'
Error: Process completed with exit code 2.

--- a/src/ports/tang_nano_4k/Makefile
+++ b/src/ports/tang_nano_4k/Makefile
@@ -33,10 +33,12 @@ FLASH_ORIGIN = 0x60000000
 FLASH_LENGTH = 4M
 else
 FLASH_ORIGIN = 0x00000000
-FLASH_LENGTH = 32K
+# The GW1NSR-4C has 128KB of address space for instruction flash (0x00000 to 0x1FFFF)
+# as per the memory map in the GW1NSR-4C datasheet.
+FLASH_LENGTH = 128K
 endif
 
-LDFLAGS += -T tang_nano_4k.ld -Wl,-Map=$@.map,--cref,--gc-sections --specs=nosys.specs -lc -lnosys
+LDFLAGS += -T tang_nano_4k.ld -Wl,-Map=$@.map,--cref,--gc-sections --specs=nano.specs --specs=nosys.specs -lc -lnosys
 LDFLAGS += -Wl,--defsym=FLASH_ORIGIN=$(FLASH_ORIGIN),--defsym=FLASH_LENGTH=$(FLASH_LENGTH)
 
 # Tune for Debugging or Optimization

--- a/src/ports/tang_nano_4k/mpconfigport.h
+++ b/src/ports/tang_nano_4k/mpconfigport.h
@@ -42,6 +42,9 @@ extern const struct _mp_obj_module_t mp_module_time;
 #define MICROPY_PY_MACHINE_SOFTSPI (1)
 #define MICROPY_PY_MACHINE_MEMX (1)
 
+#define MICROPY_PY_BUILTINS_FLOAT (0)
+#define MICROPY_PY_MATH (0)
+
 #define MICROPY_VFS (1)
 #define MICROPY_VFS_LFS2 (1)
 #define MICROPY_PY_UOS (1)

--- a/test/tang_nano_4k.repl
+++ b/test/tang_nano_4k.repl
@@ -6,9 +6,9 @@ cpu: CPU.CortexM @ sysbus
     cpuType: "cortex-m3"
     nvic: nvic
 
-// Internal Flash (User Flash) - 32KB
+// Internal Flash (Code Flash) - 128KB
 flash: Memory.MappedMemory @ sysbus 0x00000000
-    size: 0x8000
+    size: 0x20000
 
 // Internal SRAM - 22KB
 sram: Memory.MappedMemory @ sysbus 0x20000000

